Refactor main.rs

JasonLG1979 opened the pull request:
* Don't panic when parsing options. Instead list valid values and exit.
* Get rid of needless .expect in playback/src/audio_backend/mod.rs.
* Enforce reasonable ranges for option values (breaking).
* Don't evaluate options that would otherwise have no effect.
* Add pub const MIXERS to mixer/mod.rs very similar to the audio_backend's implementation. (non-breaking though)
* Use different option descriptions and error messages based on what backends are enabled at build time.
* Add a -q, --quiet option that changed the logging level to warn.
* Add a short name for every flag and option.
* Note removed options.
* Other misc cleanups.

roderickvd commented:
Just let me know when you’re ready.

JasonLG1979 commented:
> Just let me know when you’re ready.

I'm ready for your review I think? Everything functions as intended. It's just a matter of if there's anything you like changed?

roderickvd commented on src/main.rs:
```diff
@@ -525,12 +618,19 @@ fn get_setup(args: &[String]) -> Setup {
+            .map(|range| {
+                range.parse::<f64>().unwrap_or_else(|_| {
+                    error!("Invalid --{}: {}", VOLUME_RANGE, range);
+                    println!("Default: softvol - 60, alsa - what the control supports");
```
For bonus points,  you could introduce a `MIXERS` constant in `mixer/mod.rs` and iterate over the available mixers, like the `audio_backend/mod.rs` does for the backends.

JasonLG1979 replied:
Same question as above.

JasonLG1979 replied:
Why iterate though? There's only 2 options? Seems silly to iterate when the choice is binary?

JasonLG1979 replied:
Also another question is it reasonable to have a max range here? 96 would be a natural max as it's the dynamic range of 16 bit audio and probably more than enough but 100 also works because it's a nice round number.

roderickvd replied:
> Why iterate though? There's only 2 options? Seems silly to iterate when the choice is binary?

Like I said, it'd be for bonus points. I don't see any further mixers being added, but I like to do these things programatically instead of hard-coded. It's not a requirement for me, only something nice to have and I thought maybe easily copied from existing logic.



> Also another question is it reasonable to have a max range here? 96 would be a natural max as it's the dynamic range of 16 bit audio and probably more than enough but 100 also works because it's a nice round number.

100 dB is a good number. Note that this is more related to dB SPL than to DNR: https://www.dr-lex.be/info-stuff/volumecontrols.html

roderickvd commented on src/main.rs:
```diff
@@ -539,12 +639,17 @@ fn get_setup(args: &[String]) -> Setup {
-                    .expect("Invalid volume control type")
+                VolumeCtrl::from_str_with_range(volume_ctrl, volume_range).unwrap_or_else(|_| {
+                    error!("Invalid --{}: {}", VOLUME_CTRL, volume_ctrl);
+                    println!("Valid --{} values: cubic, fixed, linear, log", VOLUME_CTRL);
```
Same for bonus points.

JasonLG1979 replied:
Will do.

roderickvd commented on src/main.rs:
```diff
@@ -608,11 +720,20 @@ fn get_setup(args: &[String]) -> Setup {
+            let volume = initial_volume.parse::<u16>().unwrap_or_else(|_| {
+                error!("Invalid --{}: {}", INITIAL_VOLUME, initial_volume);
+                println!("Valid --{} values: 0-100", INITIAL_VOLUME);
+                println!("Default: 50");
```
Or the current value when the Alsa mixer is used.

JasonLG1979 replied:
This will get you the actual used backend's name. It would be trivial to then basically say: "Well you're not using the alsa backend you can't use any alsa specific functionality."

```rust
pub fn find_name(name: Option<String>) -> &'static str {
    if let Some(name) = name {
        BACKENDS
            .iter()
            .find(|backend| name == backend.0)
            .map(|backend| backend.0).unwrap_or_default()
    } else {
        BACKENDS.first().map(|backend| backend.0).unwrap_or_default()
    }
}
```

JasonLG1979 replied:
Or to make it even easier find could just also return the name along side SinkBuilder basically the whole backend tuple.

JasonLG1979 replied:
```rust
pub fn find(name: Option<String>) -> Option<&'static(&'static str, SinkBuilder)> {
    if let Some(name) = name {
        BACKENDS.iter().find(|backend| name == backend.0)
    } else {
        BACKENDS.first()
    }
}
```

JasonLG1979 replied:
I probably would rather add a function than break the API though if we go down that road.

JasonLG1979 replied:
1/2 done. I added ", or the current value when the Alsa mixer is used."

roderickvd replied:
> Same question as above as far as if it makes sense to expose alsa specific options even if it's compiled in if you're not actually using the alsa backend?

I see what you mean now. I do think it would be nice to bail out with an error. On the other hand, I think that in manu GNU tools you can specify all the parameters you want and they'll be silently ignored if they don't apply. Not to say that the GNU tools are the Golden Standard, but just as an example.

JasonLG1979 replied:
I think we should exit if the backend is not alsa but alsa volume control is selected but just warn if alsa volume modifiers are called in combo with softvol.

JasonLG1979 replied:
To clarify it's possible to use alsa hardware volume control in combination  with other backends. I'm not sure if that's a feature or a bug?

roderickvd replied:
Hmm, I think it's unintended but might be useful to some. After all, some other backend *may* in the end pipe to Also after all. Printing a warning seems a fine idea.

JasonLG1979 replied:
Let's continue this in the thread below.

roderickvd commented on src/main.rs:
```diff
@@ -663,77 +798,269 @@ fn get_setup(args: &[String]) -> Setup {
             .map(|method| {
-                NormalisationMethod::from_str(method).expect("Invalid normalisation method")
+                warn!(
+                    "--{} will be deprecated in a future release.",
```
While you're here, could you add a deprecation notice for the basic normaliser? Can be here or in `player.rs`.

JasonLG1979 replied:
Will do. I like them here if that's OK?

JasonLG1979 replied:
Done.

roderickvd commented on src/main.rs:
```diff
@@ -663,77 +798,269 @@ fn get_setup(args: &[String]) -> Setup {
-                Duration::from_millis(release.parse::<u64>().expect("Invalid release value"))
+                if !normalisation {
+                    warn!(
+                        "Without the --{} flag --{} has no effect.",
```
This warning is best also placed with the Alsa mixer options.

JasonLG1979 replied:
This doesn't have anything really to do with the alsa mixer does it?

roderickvd replied:
No, I was thinking along specifying `--alsa-xxx` options without the Alsa mixer enabled.

JasonLG1979 replied:
This a normalization limiter setting it has nothing to do with the alsa mixer.

roderickvd replied:
I know I am taking the “without x flag y has no effect” as an example. For consistency you should also do that for all the Alsa flags, that’s my point.

JasonLG1979 replied:
The problem really is that if the alsa backend is compiled in you can select a different backend (ofc if it's also compiled in) and select the alsa volume method and the alsa setting will have an effect even though you're  not using the alsa backend.

JasonLG1979 replied:
How's that? Now alsa specific options aren't even evaluated unless the alsa backend is compiled in. The user just gets a warning of  "Alsa specific options have no effect if the alsa backend is not enabled at build time." if they try to use any.

JasonLG1979 replied:
There. @roderickvd I think that about covers it as far as warnings. I paired the warnings down a little bit actually but no options that would logically do nothing, like normalization options if normalization is not enabled are not even evaluated but instead just set to their defaults.

JasonLG1979 replied:
So as a recap:
 1. We're calling the ability to use alsa volume control with other backends a feature not a bug even though it may be unintentional and treating it as such and not throwing errors in that case.
 
 2. Options that would logically have no effects are not evaluated, default values are used, but the user is warned that such options have no effect if they try to use them.
 
 3. If an option is evaluated and an error happens the user is given a useful error message that tells them the valid values for the option and the default if applicable and librespot exits with an error.
 
 4. The options and flags have been cleaned up both order and formatting wise so main looks nicer and so the help menu is prettier.
 
 5. Some other misc clean up here and there.

JasonLG1979 replied:
Ok @roderickvd I feature gated the error messages and option descriptions so for example here's what `--help` looks like with just the PulseAudio backend complied in:

```
Usage: ./librespot [options]

Options:
    -h, --help          Print this help menu.
    -v, --verbose       Enable verbose log output.
    -V, --version       Display librespot version string.
        --disable-audio-cache 
                        Disable caching of the audio data.
        --disable-credential-cache 
                        Disable caching of credentials.
        --disable-discovery 
                        Disable zeroconf discovery mode.
        --disable-gapless 
                        Disable gapless playback.
        --emit-sink-events 
                        Run PROGRAM set by `--onevent` before the sink is
                        opened and after it is closed.
        --enable-volume-normalisation 
                        Play all tracks at approximately the same apparent
                        volume.
        --autoplay      Automatically play similar songs when your music ends.
        --passthrough   Pass a raw stream to the output. Only works with the
                        pipe and subprocess backends.
    -n, --name NAME     Device name. Defaults to Librespot.
    -b, --bitrate BITRATE
                        Bitrate (kbps) {96|160|320}. Defaults to 160.
    -u, --username USERNAME
                        Username used to sign in with.
    -p, --password PASSWORD
                        Password used to sign in with.
    -c, --cache PATH    Path to a directory where files will be cached.
    -m, --mixer MIXER   Not supported by the current audio backend(s).
        --device-type TYPE
                        Displayed device type. Defaults to speaker.
        --system-cache PATH
                        Path to a directory where system files (credentials,
                        volume) will be cached. May be different from the
                        `--cache` option value.
        --cache-size-limit SIZE
                        Limits the size of the cache for audio files. It's
                        possible to use suffixes like K, M or G, e.g. 16G for
                        example.
        --onevent PROGRAM
                        Run PROGRAM when a playback event occurs.
        --proxy URL     HTTP proxy to use when connecting.
        --ap-port PORT  Connect to an AP with a specified port 1 - 65535. If
                        no AP with that port is present a fallback AP will be
                        used. Available ports are usually 80, 443 and 4070.
        --backend NAME  Audio backend to use. Use ? to list options.
        --device NAME   Not supported by the current audio backend(s).
        --format FORMAT Output format {F64|F32|S32|S24|S24_3|S16}. Defaults to
                        S16.
        --dither DITHER Specify the dither algorithm to use
                        {none|gpdf|tpdf|tpdf_hp}. Defaults to tpdf for formats
                        S16, S24, S24_3 and none for other formats.
        --mixer-name    
        --alsa-mixer-control NAME
                        Not supported by the current audio backend(s).
        --mixer-card    
        --alsa-mixer-device DEVICE
                        Not supported by the current audio backend(s).
        --mixer-index   
        --alsa-mixer-index NUMBER
                        Not supported by the current audio backend(s).
        --initial-volume VOLUME
                        Initial volume in % from 0 - 100. Defaults to 50.
        --zeroconf-port PORT
                        The port the internal server advertises over zeroconf
                        1 - 65535. Ports <= 1024 may require root privileges.
        --normalisation-method METHOD
                        Specify the normalisation method to use
                        {basic|dynamic}. Defaults to dynamic.
        --normalisation-gain-type TYPE
                        Specify the normalisation gain type to use
                        {track|album|auto}. Defaults to auto.
        --normalisation-pregain PREGAIN
                        Pregain (dB) applied by volume normalisation from
                        -10.0 to 10.0. Defaults to 0.0.
        --normalisation-threshold THRESHOLD
                        Threshold (dBFS) at which point the dynamic limiter
                        engages to prevent clipping from 0.0 to -10.0.
                        Defaults to -2.0.
        --normalisation-attack TIME
                        Attack time (ms) in which the dynamic limiter reduces
                        gain from 1 to 500. Defaults to 5.
        --normalisation-release TIME
                        Release or decay time (ms) in which the dynamic
                        limiter restores gain from 1 to 1000. Defaults to 100.
        --normalisation-knee KNEE
                        Knee steepness of the dynamic limiter from 0.0 to 2.0.
                        Defaults to 1.0.
        --volume-ctrl VOLUME_CTRL
                        Volume control scale type {cubic|fixed|linear|log}.
                        Defaults to log.
        --volume-range RANGE
                        Range of the volume control (dB) from 0.0 to 100.0.
                        Defaults to 60.0.
```

JasonLG1979 replied:
Notice no mention of softvol vs alsa and the alsa specific options say `Not supported by the current audio backend(s).`

roderickvd replied:
Does look much nicer!

JasonLG1979 commented on src/main.rs:
```diff
@@ -197,8 +197,8 @@ fn get_setup(args: &[String]) -> Setup {
     const BACKEND: &str = "backend";
-    const BITRATE: &str = "b";
-    const CACHE: &str = "c";
+    const BITRATE: &str = "bitrate";
+    const CACHE: &str = "cache";
```
These were the only ones that were the short version of the options/flags (b, c) I changed them for the sake of consistency. I also changed help for the same reason. I fixed them in 

Old (inconsistent):
https://github.com/librespot-org/librespot/blob/dev/src/main.rs#L200-L201
https://github.com/librespot-org/librespot/blob/dev/src/main.rs#L213
https://github.com/librespot-org/librespot/blob/dev/src/main.rs#L240-L248
https://github.com/librespot-org/librespot/blob/dev/src/main.rs#L263-L267

New (consistent):
https://github.com/librespot-org/librespot/blob/f353759b31654388a884e0b5b063f0b387428afd/src/main.rs#L240-L243
https://github.com/librespot-org/librespot/blob/f353759b31654388a884e0b5b063f0b387428afd/src/main.rs#L301-L305
https://github.com/librespot-org/librespot/blob/f353759b31654388a884e0b5b063f0b387428afd/src/main.rs#L319-L323

You'll also notice that I changed the order to all flags and then options and each section starts with the ones with short versions so that the help menu looks nicer IMHO.

JasonLG1979 commented on src/main.rs:
```diff
@@ -476,13 +553,23 @@ fn get_setup(args: &[String]) -> Setup {
+            MIXER_TYPE,
+            matches.opt_str(MIXER_TYPE).unwrap_or_default()
+        );
+        println!("Valid --{} values: alsa, softvol", MIXER_TYPE);
```
Does it even make sense for alsa specific options to work with different backends even if the alsa backend may be compiled in?

roderickvd replied:
Not sure I understand your question. Of course the Alsa options only apply to the Alsa backend or mixer?

JasonLG1979 commented:
Squashed and ready for another round of reviews...

JasonLG1979 commented:
@roderickvd There I've made `mixer::find` function like `audio_backend::find` except I didn't change the name arg type. It's still `Option<&str>` to prevent API breakage.

roderickvd commented:
Nice work man. You have now officially reached "Large diffs are not rendered by default" status on GitHub 😆 
I'll take a better look at it the coming few days!

JasonLG1979 commented:
> Nice work man. You have now officially reached "Large diffs are not rendered by default" status on GitHub 
> I'll take a better look at it the coming few days!

I was there a couple days ago,lol!!! It looks like a lot and it kinda is but it's also really not. It's mostly polish.

michaelherger commented:
> It looks like a lot and it kinda is but it's also really not. It's mostly polish.

Heh... and still this will be a tough one for me to merge: most of my customizations are in `main.rs`: I've added a few params, removed some others. Exactly the area you changed so much here 😀. I'm not complaining, I'm thankful somebody is actively maintaining this project.

JasonLG1979 commented:
> Heh... and still this will be a tough one for me to merge: most of my customizations are in `main.rs`: I've added a few params, removed some others. Exactly the area you changed so much here . I'm not complaining, I'm thankful somebody is actively maintaining this project.

That would be expected.  Main is really just glue code that helps hold all the librespot libs together. If you want to maintain a divergent fork of the librespot binary you'd be expected to maintain your own patches/changes. Make your own glue so to speak. It would probably be more maintainable in the long run to write your own application that just uses the librespot libs.

JasonLG1979 commented:
I also noticed that you're using the basic normalization mode. The basic mode will be deprecated in the near future. If the dynamic doesn't work for you for whatever reason you might want to file a bug/figure out why or basic mode will be another patch you'll need to carry.

michaelherger commented:
> That would be expected. Main is really just glue code that helps hold all the librespot libs together. If you want to maintain a divergent fork of the librespot binary you'd be expected to maintain your own patches/changes. Make your own glue so to speak. It would probably be more maintainable in the long run to write your own application that just uses the librespot libs.

I thought you might suggest this. And I actually did this. But it proved to be harder for me (as a not fluent in Rust developer) to pick the relevant changes from librespot to translate them to whatever I would need. Dealing with conflicts in the params part I can handle.

Thanks for the hint about the normalization mode. I'll go with whatever is the default behavior. I don't really care about it, as I personally don't use it 😀.

JasonLG1979 commented:
> I thought you might suggest this. And I actually did this. But it proved to be harder for me (as a not fluent in Rust developer) to pick the relevant changes from librespot to translate them to whatever I would need. Dealing with conflicts in the params part I can handle.

For now maybe. Eventually as librespot moves forward and drifts away from your fork you're going to have problems. Especially with the rate that development has increased in librespot recently. Your fork is already +26/-57 commits. That gap is only going to get wider. If I were you I'd pull the librespot bits in as dependencies like what spotifyd does and keep just the glue code like main.

michaelherger commented:
My dev branch is "only" +33/-2: https://github.com/michaelherger/librespot/tree/spotty-dev. Most changes usually are straight forward merges. As would be the missing 2. I'm waiting for this PR to be merged before I sync up again.

I've done the switch from using my own implementation pulling in librespot as dependencies to using a fork before. The way back shouldn't be harder. And until now keeping my fork up to date has been much easier than keeping a standalone version in sync. Because most of the changes usually merge easily. Conflicting wording in a description even I can handle.

As I said: I'm not complaining at all. I only was kind of joking about how a seemingly simple cleanup of runtime options descriptions messed with my customizing this particular part of the code.

JasonLG1979 commented:
@michaelherger I did a compare of your spotty dev branch to your master and the only differences are Logitech Media Server specific changes. You haven't change any of the core librespot stuff at all that I saw. You are absolutely wasting your time and effort maintaining a librespot fork with a divergent branch. A lot of it could just be added to the wiki and the contrib section of this repo and/or (I know I keep saying this,lol!!!) you could just maintain the handful of files that are changed/new and pull librespot in as a lib.

roderickvd commented:
You just keep on pushing and pushing 😆 again let me know when you're ready for review.

JasonLG1979 commented:
> You just keep on pushing and pushing 😆 again let me know when you're ready for review.

Just tweaking. I'll stop. Jump in there and give me your worst,lol!!!

JasonLG1979 commented:
@roderickvd I have a couple questions/ideas?

1. `librespot` defaults to the info log level which is fine and all but a little bit spammy if you're running it as service. It generates A LOT of logs for basically no reason. It would be nice to have a `-q`, `--quiet` option that changed the logging level to warn to generate basically no logs unless there are actual problems.

2. It would also be really nice if every option/flag had a short version. Most currently do not. With all the current options and flags an arg string can get pretty crazy long pretty quick and some of the options have really long names. Mine for example is 345 chars long start to finish granted I tweak just about everything but still.

JasonLG1979 commented:
On an unrelated note I just update to the latest version of Raspberry Pi OS (lite) that's based on Debian Bullseye and everything seem to work perfectly.

roderickvd commented:
> 1. `librespot` defaults to the info log level which is fine and all but a little bit spammy if you're running it as service. It generates A LOT of logs for basically no reason. It would be nice to have a `-q`, `--quiet` option that changed the logging level to warn to generate basically no logs unless there are actual problems.

Principally sounds good to me. Could howver already be done by piping to `/dev/null`.

> 2. It would also be really nice if every option/flag had a short version. Most currently do not. With all the current options and flags an arg string can get pretty crazy long pretty quick and some of the options have really long names. Mine for example is 345 chars long start to finish granted I tweak just about everything but still.

Fine by me.

JasonLG1979 commented:
> Principally sounds good to me. Could howver already be done by piping to `/dev/null`.

That's not really something you'd do with a system service unless it was spamming your logs... *cough, cough... like librespot does...* librespot 's logging is crazy verbose by default. Most apps default to warning level and only go to info in verbose mode and go to trace/debug level in debug mode.

> Fine by me.

Ok. We have so many options we almost run though the English alphabet lower case and upper case since short names in `getopts` can only be 1 char. It starts to get pretty arbitrary real quick as far as the short name not having any connection to the long name.

roderickvd commented on src/main.rs:
```diff
@@ -450,39 +605,122 @@ fn get_setup(args: &[String]) -> Setup {
 
-    let verbose = matches.opt_present(VERBOSE);
-    setup_logging(verbose);
+    setup_logging(matches.opt_present(QUIET), matches.opt_present(VERBOSE));
```
Quiet and verbose should be mutually exclusive (not allowed on the command line).

JasonLG1979 replied:
Can do.

JasonLG1979 replied:
I will make a note to the user with a warning. In my mind if both are given verbose should trump quiet.

JasonLG1979 replied:
Done.

roderickvd commented on src/main.rs:
```diff
@@ -67,9 +68,9 @@ fn setup_logging(verbose: bool) {
+    println!("Available backends: ");
     for (&(name, _), idx) in BACKENDS.iter().zip(0..) {
-        if idx == 0 {
+        if matches!(idx, 0) {
```
Is this something that clippy wanted? You're not matching against an enum are you?

JasonLG1979 replied:
I just went thought and changed all == to matches that I could without major changes I suppose == still makes logical sense for numbers but for non numbers I think I prefer matches. I can change back to == for number type equality if you like.

JasonLG1979 replied:
I changed all number comparisons back to ==.

roderickvd commented on src/main.rs:
```diff
@@ -236,85 +238,234 @@ fn get_setup(args: &[String]) -> Setup {
+    #[cfg(feature = "alsa-backend")]
+    const MIXER_TYPE_DESC: &str = "Mixer to use {alsa|softvol}. Defaults to softvol.";
+    #[cfg(not(feature = "alsa-backend"))]
+    const MIXER_TYPE_DESC: &str = "Not supported by the current audio backend(s).";
```
Do you think "current" is concise enough? I'm wondering if it should be "compiled" or "included" or something like that.

JasonLG1979 replied:
I like included it's less technical sounding than compiled and I think easier for average Joe to understand. I can change it.

JasonLG1979 replied:
Done.

roderickvd commented on src/main.rs:
```diff
@@ -494,63 +732,121 @@ fn get_setup(args: &[String]) -> Setup {
-                index
-                    .parse::<u32>()
-                    .expect("Mixer index is not a valid number")
+                warn!("`--mixer-index` is deprecated and will be removed in a future release.");
```
I'm thinking of this now, but we deprecated these in 0.3.0 did we not? If so I'd be OK with removing them now, since the next release will be 0.4.0 due to API breakage.

JasonLG1979 replied:
Can do.

JasonLG1979 replied:
Done.

roderickvd commented on src/main.rs:
```diff
@@ -494,63 +732,121 @@ fn get_setup(args: &[String]) -> Setup {
+                    })
                 })
-                .unwrap_or(0),
+                .unwrap_or_else(|| MixerConfig::default().index),
```
`unwrap_or` might suffice here, because that default index is a constant.

JasonLG1979 replied:
It may not always be in the future though?

JasonLG1979 replied:
In the future who knows the logic to find `MixerConfig::default().index` might change and maybe it won't always be zero? I don't know. You're the mixer guy. Do you foresee that being a possibility?

JasonLG1979 replied:
I also dried up the `ConnectConfig::default()` stuff.

JasonLG1979 replied:
`MixerConfig::default()` is also dried up too.

roderickvd commented on src/main.rs:
```diff
@@ -663,89 +1029,343 @@ fn get_setup(args: &[String]) -> Setup {
+                }
+            }
+
+            normalisation_method = PlayerConfig::default().normalisation_method;
```
You could DRY this up by doing a `let default_config = PlayerConfig::default()` first.

JasonLG1979 replied:
Can do.

JasonLG1979 replied:
Done.

roderickvd reviewed (commented):
This is looking excellent. Some final minor points from my code review. I haven't taken the time to test every combination...

JasonLG1979 commented:
Ok @roderickvd I'm ready for a new round. I have deleted my local clones so I can't fiddle with  it :laughing: so I won't push anything else unless you'd like changes made.

roderickvd commented:
It's a bit hard for me to see what the changes are between your latest commit and my previous review, and GitHub isn't helping. Do you have a link to a diff between the two?

I think it's easiest to save force-pushing for those occasions where a commit was in error, or after squashing just before merging.

JasonLG1979 commented:
> It's a bit hard for me to see what the changes are between your latest commit and my previous review, and GitHub isn't helping. Do you have a link to a diff between the two?

No you'll have to either trust me or re-review it.

> I think it's easiest to save force-pushing for those occasions where a commit was in error, or after squashing just before merging.

I squashed and force pushed to keep the history clean. If you look at my previous merged PR's there's a lot of noise in there. Github's squash button keeps all that noise. You have to do it by hand to clean out the noise.

From now on I will wait to squash until right before you're ready to merge.

JasonLG1979 commented:
I was made a maintainer of Raspotify. Waiting on you to review and merge this @roderickvd so I can include the `--quiet ` and the non-explodey arg parsing fails for the sweaty masses,lol!!!

roderickvd commented on src/main.rs:
```diff
@@ -450,110 +593,213 @@ fn get_setup(args: &[String]) -> Setup {
     let mixer_config = {
-        let mixer_device = match matches.opt_str("mixer-card") {
-            Some(card) => {
-                warn!("--mixer-card is deprecated and will be removed in a future release.");
```
We should document in the changelog which previously deprecated features have now been removed.

JasonLG1979 replied:
Can do.

JasonLG1979 replied:
Done.

roderickvd commented on src/main.rs:
```diff
@@ -450,110 +593,213 @@ fn get_setup(args: &[String]) -> Setup {
+        });
+
+        #[cfg(not(feature = "alsa-backend"))]
+        let device = mixer_default_config.device;
```
Sorry if this was in before and I missed it in my earlier review. I'm thinking you could do `let out device = mixer_default_config.device` before the `#[cfg(feature = "alsa-backend")]` pragma and then only override device when a mixer supports that. It might cut down on a couple of pragma's, and make it easier to add a new mixer in the future (you'd only the need `#[cfg(feature = "new-mixer-thingy)]`

JasonLG1979 replied:
What new mixer devices do you see adding in the future? Hardware volume control on Mac and Windows or something?

roderickvd replied:
At this point I don't see any, although apps downstream might add their own. Still, my suggestion might cut down a few lines of code.

JasonLG1979 replied:
> At this point I don't see any, although apps downstream might add their own. Still, my suggestion might cut down a few lines of code.

OK sure I can do that but downstream apps wouldn't and shouldn't use our main.rs. They should use their own main and use the librespot crates and libs.

JasonLG1979 replied:
```rust
        let mut device = mixer_default_config.device;

        #[cfg(feature = "alsa-backend")]
        device = matches.opt_str(ALSA_MIXER_DEVICE).unwrap_or_else(|| {
            if let Some(ref device_name) = device {
                device_name.to_string()
            } else {
                mixer_default_config.device.clone()
            }
        });
```

Does not work.

```
error[E0658]: attributes on expressions are experimental
   --> src/main.rs:715:9
    |
715 |         #[cfg(feature = "alsa-backend")]
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: see issue #15701 <https://github.com/rust-lang/rust/issues/15701> for more information

error: removing an expression is not supported in this position
   --> src/main.rs:715:9
    |
715 |         #[cfg(feature = "alsa-backend")]
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0308]: mismatched types
   --> src/main.rs:717:20
    |
717 |             if let Some(ref device_name) = device {
    |                    ^^^^^^^^^^^^^^^^^^^^^   ------ this expression has type `std::string::String`
    |                    |
    |                    expected struct `std::string::String`, found enum `std::option::Option`
    |
    = note: expected struct `std::string::String`
                 found enum `std::option::Option<_>`

Some errors have detailed explanations: E0308, E0658.
For more information about an error, try `rustc --explain E0308`.
```

JasonLG1979 replied:
I also think that it's more explicit the way it is.

roderickvd replied:
Did you forget a `let` though after that pragma?
If it takes too much time then don't bother.

JasonLG1979 replied:
Derp...

JasonLG1979 replied:
```rust
        let device = mixer_default_config.device;

        #[cfg(feature = "alsa-backend")]
        let device = matches.opt_str(ALSA_MIXER_DEVICE).unwrap_or_else(|| {
            if let Some(ref device_name) = device {
                device_name.to_string()
            } else {
                mixer_default_config.device.clone()
            }
        });
```

Doesn't work either.

```
   --> src/main.rs:717:20
    |
717 |             if let Some(ref device_name) = device {
    |                    ^^^^^^^^^^^^^^^^^^^^^   ------ this expression has type `std::string::String`
    |                    |
    |                    expected struct `std::string::String`, found enum `std::option::Option`
    |
    = note: expected struct `std::string::String`
                 found enum `std::option::Option<_>`
```

Because:

```rust
        let device = mixer_default_config.device; #<== Is a String.

        #[cfg(feature = "alsa-backend")]
        let device = matches.opt_str(ALSA_MIXER_DEVICE).unwrap_or_else(|| {
            if let Some(ref device_name) = device { #<== This is testing against the 1st "let device" not if device_name is Some.
                device_name.to_string()
            } else {
                mixer_default_config.device.clone()
            }
        });
```

JasonLG1979 replied:
Can we agree to cross this bridge when/if it comes. I'd really like to get this merged.

JasonLG1979 replied:
Let me know when you want me to squash this.

roderickvd replied:
Yes I agree. You can squash if you want to or we can have GitHub do a squash & merge. Let's merge this! 👍

JasonLG1979 replied:
I'll squash it. Give me a sec.

JasonLG1979 replied:
All squished and ready as soon as it passes CI.

roderickvd commented on src/main.rs:
```diff
@@ -450,110 +593,213 @@ fn get_setup(args: &[String]) -> Setup {
+                    exit(1);
+                });
+
+                if !(0.0..=100.0).contains(&range) {
```
Final nit and again sorry if I could have mentioned this before. If you'd make these into constants like `MAX_VOLUME_RANGE` then you could re-use those in the `on_error` lines and not repeat yourself.

JasonLG1979 replied:
Can do. I'll do that for all the ranges.

JasonLG1979 replied:
Done.

roderickvd commented on src/main.rs:
```diff
@@ -647,13 +878,131 @@ fn get_setup(args: &[String]) -> Setup {
+                    );
+                    #[cfg(feature = "alsa-backend")]
+                    println!(
+                        "Default: {}, or the current value when the Alsa mixer is used.",
```
Small nit, we're using "alsa" and "Alsa" interchangeably in various lines.

JasonLG1979 replied:
I can change everything to "alsa"

JasonLG1979 replied:
Done.

roderickvd reviewed (commented):
Final nits! This is shaping up to be a major improvement on the way to a robust command-line application, thanks.

JasonLG1979 commented:
Happy day. CI passed!!!

roderickvd commented:
Great work man 🍻

JasonLG1979 commented:
Thanks. Now the plan is to do a release of Raspotify that tracks dev and then we can use Raspotify users to shake out bugs and every so often I'll do future release of Raspotify as snapshots of dev. So in effect Raspotify will be a rolling unstable-ish librespot. I'll try to filter the noise from Raspotify. I'm not really interested in debugging someone's network issues or really dealing with docker containers and all that. I'm just looking for real bugs.

roderickvd commented:
If we ever get to the point of separating `librespot` as a library and `librespotd` as the Connect daemon, then we should look at https://github.com/clap-rs/clap.

Note that it's not at all high on my list, so this is just for future reference.

JasonLG1979 commented:
> If we ever get to the point of separating `librespot` as a library and `librespotd` as the Connect daemon, then we should look at https://github.com/clap-rs/clap.
> 
> Note that it's not at all high on my list, so this is just for future reference.

I'd personally like a proper config file parser. So you could just so `librespot --config-file /foo/bar`.

JasonLG1979 commented:
At this point really librespot and librespotd are starting to look like the same thing. As in librespot is currently able to be a 'd' as evidenced by Raspotify which is literally just an unaltered librespot binary, a systemd unit file and an environment file.
